src/lib/TopLevel.hs

```diff
@@ -604,18 +606,42 @@ evalBlock typed = do
   applyRecon recon result
 {-# SCC evalBlock #-}
 
-evalSpecializations :: (Topper m, Mut n) => [AtomName n] -> m n ()
-evalSpecializations vs = forM_ vs \v -> do
-  lookupAtomName v >>= \case
-    TopFunBound _ (SpecializedTopFun (AppSpecialization _ _)) ->
-      queryImpCache v >>= \case
-        Nothing -> compileTopLevelFun v
-        Just _ -> return ()
-    TopFunBound _ (SpecializedTopFun (IxMethodSpecialization _ _)) -> do
-      queryIxLoweredCache v >>= \case
-        Nothing -> lowerIxMethod v
-        Just _ -> return ()
-    _ -> return ()
+evalSpecializations :: (Topper m, Mut n) => [AtomName n] -> [SpecDictName n] -> m n ()
+evalSpecializations vs sdVs = do
+  forM_ vs \v -> do
+    lookupAtomName v >>= \case
+      TopFunBound _ (SpecializedTopFun (AppSpecialization _ _)) ->
+        queryImpCache v >>= \case
+          Nothing -> compileTopLevelFun v
+          Just _ -> return ()
+      _ -> return ()
+  forM_ sdVs \d -> do
+    SpecializedDictBinding (SpecializedDict absDict@(Abs bs _) _) <- lookupEnv d
```

Comment:
Might be a good idea to assert `Nothing` here as well, just to make sure we're not wasting work.